akka-remote/src/main/scala/akka/remote/artery/SendQueue.scala

```diff
@@ -0,0 +1,112 @@
+/**
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.remote.artery
+
+import java.util.Queue
+import akka.stream.stage.GraphStage
+import akka.stream.stage.OutHandler
+import akka.stream.Attributes
+import akka.stream.Outlet
+import akka.stream.SourceShape
+import akka.stream.stage.GraphStageLogic
+import org.agrona.concurrent.ManyToOneConcurrentArrayQueue
+import akka.stream.stage.GraphStageWithMaterializedValue
+import org.agrona.concurrent.ManyToOneConcurrentLinkedQueueTail
+import org.agrona.concurrent.ManyToOneConcurrentLinkedQueue
+import java.util.concurrent.atomic.AtomicInteger
+import scala.annotation.tailrec
+
+/**
+ * INTERNAL API
+ */
+private[remote] object SendQueue {
+  trait QueueValue[T] {
+    def inject(queue: Queue[T]): Unit
+    def offer(message: T): Boolean
+  }
+
+  private trait WakeupSignal {
+    def wakeup(): Unit
+  }
+}
+
+/**
+ * INTERNAL API
+ */
+private[remote] final class SendQueue[T] extends GraphStageWithMaterializedValue[SourceShape[T], SendQueue.QueueValue[T]] {
+  import SendQueue._
+
+  val out: Outlet[T] = Outlet("SendQueue.out")
+  override val shape: SourceShape[T] = SourceShape(out)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, QueueValue[T]) = {
+    @volatile var needWakeup = false
+    @volatile var queue: Queue[T] = null
```

Comment:
Use a Promise/Future then. This is not a performance critical path since the injection happens only once. OTOH the queue not being volatile gives more options for JIT (for example keep the reference in register).